# Linguistic Style Analysis Across Classic Literature

Ingrid Foslien | September 2026

This notebook explores differences in vocabulary richness, sentence
structure, and word usage across seven classic novels using NLTK's
Gutenberg corpus. It compares authors including Jane Austen, Herman
Melville, Lewis Carroll, and G.K. Chesterton, and includes an early
prototype for ranking books by overall style similarity. It's a step
toward a full interactive dashboard comparing linguistic style across
authors and works.

In [1]:
import nltk

nltk.download('gutenberg')
nltk.download('punkt_tab')

from nltk.corpus import gutenberg

[nltk_data] Downloading package gutenberg to
[nltk_data]     /Users/iafoslien/nltk_data...
[nltk_data]   Package gutenberg is already up-to-date!
[nltk_data] Downloading package punkt_tab to
[nltk_data]     /Users/iafoslien/nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!


In [10]:
# Compute basic linguistic stats for a single book
import textstat

def analyze_book(book, ttr_sample_size=10000):
    words = gutenberg.words(book)
    sents = gutenberg.sents(book)
    vocab = set(w.lower() for w in words if w.isalpha())
    raw_text = gutenberg.raw(book)

    word_count = len(words)
    sentence_count = len(sents)
    vocab_size = len(vocab)

    alpha_words = [w.lower() for w in words if w.isalpha()]
    sample = alpha_words[:ttr_sample_size]
    sample_vocab = set(sample)

    return {
        'book': book,
        'word_count': word_count,
        'sentence_count': sentence_count,
        'vocab_size': vocab_size,
        'type_token_ratio_raw': round(vocab_size / word_count, 4),
        'type_token_ratio_sample': round(len(sample_vocab) / len(sample), 4),
        'avg_sentence_length': round(word_count / sentence_count, 2),
        # higher score means easier to read
        'flesch_reading_ease': round(textstat.flesch_reading_ease(raw_text), 2),
        # roughly the US school grade level needed to understand the text
        'flesch_kincaid_grade': round(textstat.flesch_kincaid_grade(raw_text), 2)
    }

In [24]:
# Run the analysis on each book and collect the results
books = [
    'austen-emma.txt',
    'austen-persuasion.txt',
    'austen-sense.txt',
    'melville-moby_dick.txt',
    'carroll-alice.txt',
    'chesterton-ball.txt',
    'chesterton-thursday.txt'
]
results = [analyze_book(book) for book in books]
results

[{'book': 'austen-emma.txt',
  'word_count': 192427,
  'sentence_count': 7752,
  'vocab_size': 7079,
  'type_token_ratio_raw': 0.0368,
  'type_token_ratio_sample': 0.1762,
  'avg_sentence_length': 24.82,
  'flesch_reading_ease': 69.29,
  'flesch_kincaid_grade': 7.8},
 {'book': 'austen-persuasion.txt',
  'word_count': 98171,
  'sentence_count': 3747,
  'vocab_size': 5739,
  'type_token_ratio_raw': 0.0585,
  'type_token_ratio_sample': 0.2011,
  'avg_sentence_length': 26.2,
  'flesch_reading_ease': 62.29,
  'flesch_kincaid_grade': 10.43},
 {'book': 'austen-sense.txt',
  'word_count': 141576,
  'sentence_count': 4999,
  'vocab_size': 6283,
  'type_token_ratio_raw': 0.0444,
  'type_token_ratio_sample': 0.1786,
  'avg_sentence_length': 28.32,
  'flesch_reading_ease': 61.78,
  'flesch_kincaid_grade': 10.06},
 {'book': 'melville-moby_dick.txt',
  'word_count': 260819,
  'sentence_count': 10059,
  'vocab_size': 16948,
  'type_token_ratio_raw': 0.065,
  'type_token_ratio_sample': 0.2816,
  'avg_

Vocabulary richness clusters cleanly by author, the three Austen
novels sit close together (0.18 to 0.20), separate from Moby Dick
(0.28) and Alice (0.15), and Chesterton's two novels also land near
each other (0.22 each).

Readability tells a different story. Persuasion and Sense and
Sensibility cluster tightly, but Emma reads noticeably easier than
Austen's other two novels, closer to Alice and Chesterton than to her
own other work. Authorial consistency holds for vocabulary but not
for readability, at least not here.

In [6]:
pip install pandas plotly

  Using cached pandas-3.0.5-cp311-cp311-macosx_11_0_arm64.whl.metadata (79 kB)
Using cached pandas-3.0.5-cp311-cp311-macosx_11_0_arm64.whl (10.0 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.1/9.1 MB 7.4 MB/s  0:00:01 eta 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.5/5.5 MB 7.0 MB/s  0:00:00 eta 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4/4 [pandas]2m3/4 [pandas]s]
Note: you may need to restart the kernel to use updated packages.


In [25]:
# turn the results list into a table for easier viewing and plotting
import pandas as pd

df = pd.DataFrame(results)
df

,book,word_count,sentence_count,vocab_size,type_token_ratio_raw,type_token_ratio_sample,avg_sentence_length,flesch_reading_ease,flesch_kincaid_grade
0,austen-emma.txt,192427,7752,7079,0.0368,0.1762,24.82,69.29,7.80
1,austen-persuasion.txt,98171,3747,5739,0.0585,0.2011,26.20,62.29,10.43
2,austen-sense.txt,141576,4999,6283,0.0444,0.1786,28.32,61.78,10.06
3,melville-moby_dick.txt,260819,10059,16948,0.0650,0.2816,25.93,64.46,10.00
4,carroll-alice.txt,34110,1703,2569,0.0753,0.1490,20.03,79.65,6.52
5,chesterton-ball.txt,96996,4779,8157,0.0841,0.2210,20.30,70.51,7.92
6,chesterton-thursday.txt,69213,3742,6300,0.0910,0.2180,18.50,73.36,7.15


### Comparing vocabulary richness fairly

The chart below shows type token ratio for each book using the same
10,000 word sample size, so all seven books are compared on equal
footing rather than penalizing the shorter ones.

In [26]:
# compare vocabulary richness across books
import plotly.express as px

fig = px.bar(
    df,
    x='book',
    y='type_token_ratio_sample',
    title='Vocabulary Richness by Book (10,000 word sample)',
    labels={'book': 'Book', 'type_token_ratio_sample': 'Type Token Ratio'},
    color_discrete_sequence=['#2a78d6']
)
fig.update_layout(template='plotly_white', showlegend=False)
fig.show()

Moby Dick has the richest vocabulary of all seven books,
consistent with its reputation for dense, varied language. Alice in
Wonderland is the lowest, a more sensible result than the raw
ratio gave, since that number was skewed by Alice being the shortest
book. The three Austen novels cluster closely together in the middle
of the range, and Chesterton's two novels land close to each other as
well, suggesting each author keeps a fairly consistent vocabulary
range across their own work.

In [27]:
pip install textstat

Note: you may need to restart the kernel to use updated packages.


### Comparing reading difficulty by book

In [28]:
# Compare readability across books
fig2 = px.bar(
    df,
    x='book',
    y='flesch_kincaid_grade',
    title='Reading Difficulty by Book (Flesch-Kincaid Grade Level)',
    labels={'book': 'Book', 'flesch_kincaid_grade': 'Grade Level'},
    color_discrete_sequence=['#2a78d6']
)
fig2.update_layout(template='plotly_white', showlegend=False)
fig2.show()

Persuasion has the highest grade level
(10.43), just above Sense and Sensibility (10.06) and Moby Dick
(10.0), all three sitting close together. Alice in Wonderland remains
the easiest by a clear margin (grade 6.52). Emma stands out within
Austen's own novels, reading noticeably easier (grade 7.8) than
Persuasion or Sense and Sensibility.

### Most frequent words by book

Looking at each book's most common words after filtering out stopwords
shows the words that carry the most meaningful content, things like
character names, objects, and themes, rather than function words like
"the" or "and".

In [14]:
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to
[nltk_data]     /Users/iafoslien/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

In [15]:
# find the most common non stopword words in a book
from nltk.corpus import stopwords
from nltk.probability import FreqDist

stop_words = set(stopwords.words('english'))

def top_words(book, n=10):
    words = gutenberg.words(book)
    filtered = [w.lower() for w in words if w.isalpha() and w.lower() not in stop_words]
    freq = FreqDist(filtered)
    return freq.most_common(n)

In [29]:
for book in books:
    print(book)
    for word, count in top_words(book):
        print(word, count)
    print()

austen-emma.txt
mr 1153
emma 865
could 837
would 820
mrs 699
miss 599
must 567
harriet 506
much 486
said 484

austen-persuasion.txt
anne 497
could 451
would 355
captain 303
mrs 291
elliot 289
mr 256
one 238
must 228
wentworth 218

austen-sense.txt
elinor 685
could 578
marianne 566
mrs 530
would 515
said 397
every 377
one 331
much 290
must 283

melville-moby_dick.txt
whale 1226
one 921
like 647
upon 566
man 527
ship 518
ahab 511
ye 472
sea 455
old 450

carroll-alice.txt
said 462
alice 398
little 128
one 104
know 88
like 85
would 83
went 83
could 77
queen 75

chesterton-ball.txt
said 652
turnbull 545
macian 426
man 338
like 326
one 316
two 177
us 162
quite 144
evan 141

chesterton-thursday.txt
syme 519
said 507
man 276
like 264
one 199
could 151
professor 149
would 120
face 119
gregory 112



The pattern holds across all seven books. Character names dominate
each list, and the specific names reveal something about each book's
structure. Persuasion surfaces both Anne Elliot and Captain Wentworth,
along with the Elliot family name, while Sense and Sensibility
surfaces both Elinor and Marianne, its two sister protagonists,
reflecting the novel's dual heroine structure. Chesterton's two novels
each center on a pair of characters too, Turnbull and MacIan in The
Ball and the Cross, and Syme, Professor, and Gregory in The Man Who
Was Thursday.

A clear genre difference also stands out. Austen's three novels all
surface formal social titles (mr, mrs, miss) among their top words,
fitting their drawing room social settings, while neither Chesterton
novel does, since they're adventure and philosophical thrillers
written decades later in a different social register.

### Sentiment across each story

Sentiment scores come from VADER, a tool built and tuned for short,
informal text like social media posts, not classic literature, so
exact values should be read loosely. The more meaningful signal here
is the shape of each line, whether tone rises, dips, or resolves by
the end, rather than precise numbers.

In [17]:
nltk.download('vader_lexicon')

[nltk_data] Downloading package vader_lexicon to
[nltk_data]     /Users/iafoslien/nltk_data...


True

In [30]:
# score sentiment across a book by splitting it into equal chunks
from nltk.sentiment import SentimentIntensityAnalyzer

sia = SentimentIntensityAnalyzer()

# score sentiment sentence by sentence, then average within each bin
def sentiment_arc(book, num_bins=20):
    sents = gutenberg.sents(book)
    sentence_scores = []
    for sent in sents:
        text = ' '.join(sent)
        score = sia.polarity_scores(text)['compound']
        sentence_scores.append(score)

    bin_size = len(sentence_scores) // num_bins
    binned_scores = []
    for i in range(num_bins):
        start = i * bin_size
        end = start + bin_size
        bin_scores = sentence_scores[start:end]
        binned_scores.append(sum(bin_scores) / len(bin_scores))
    return binned_scores

In [31]:
arc_rows = []
for book in books:
    scores = sentiment_arc(book)
    for i, score in enumerate(scores):
        arc_rows.append({'book': book, 'chunk': i + 1, 'sentiment': score})

arc_df = pd.DataFrame(arc_rows)

In [33]:
fig3 = px.line(
    arc_df,
    x='chunk',
    y='sentiment',
    color='book',
    title='Sentiment Across the Story by Book',
    labels={'chunk': 'Story Progress (chunk)', 'sentiment': 'Sentiment Score'},
    color_discrete_sequence=['#2a78d6', '#eb6834', '#1baf7a', '#eda100', '#e87ba4', '#008300', '#4a3aa7']
)
fig3.update_layout(template='plotly_white')
fig3.show()

All three Austen novels sit distinctly higher than the other four
books for nearly the entire story, staying consistently positive
between about 0.1 and 0.25. Moby Dick, Alice in Wonderland, and both
Chesterton novels cluster together in a separate lower band close to
neutral, occasionally dipping slightly negative. This is the clearest
authorial pattern found across any of the metrics so far, sentiment
tone separates Austen from the other authors more cleanly than
vocabulary richness or readability did.

### Consolidating everything into one profile table

In [34]:
# summarize each book's sentiment arc into one average and one volatility score
sentiment_summary = arc_df.groupby('book')['sentiment'].agg(['mean', 'std']).reset_index()
sentiment_summary.columns = ['book', 'avg_sentiment', 'sentiment_volatility']

df = df.merge(sentiment_summary, on='book')
df

,book,word_count,sentence_count,vocab_size,type_token_ratio_raw,type_token_ratio_sample,avg_sentence_length,flesch_reading_ease,flesch_kincaid_grade,avg_sentiment,sentiment_volatility
0,austen-emma.txt,192427,7752,7079,0.0368,0.1762,24.82,69.29,7.80,0.180160,0.046418
1,austen-persuasion.txt,98171,3747,5739,0.0585,0.2011,26.20,62.29,10.43,0.176633,0.042348
2,austen-sense.txt,141576,4999,6283,0.0444,0.1786,28.32,61.78,10.06,0.167520,0.068504
3,melville-moby_dick.txt,260819,10059,16948,0.0650,0.2816,25.93,64.46,10.00,0.026893,0.040102
4,carroll-alice.txt,34110,1703,2569,0.0753,0.1490,20.03,79.65,6.52,0.029646,0.038092
5,chesterton-ball.txt,96996,4779,8157,0.0841,0.2210,20.30,70.51,7.92,0.025679,0.037847
6,chesterton-thursday.txt,69213,3742,6300,0.0910,0.2180,18.50,73.36,7.15,0.023991,0.046561


Sentiment volatility reveals a different pattern than average tone
did. Sense and Sensibility stands out with clearly the highest
volatility (0.0685), its tone swings more dramatically across the
story than any other book, even though its average sentiment sits
close to Austen's other two novels. The remaining six books cluster
much closer together, between about 0.038 and 0.047, regardless of
whether they belong to the high sentiment Austen group or the lower
sentiment group. Volatility does not follow the same author based
split that average sentiment did.

### Finding similar books by style

Using the metrics gathered so far, vocabulary richness, sentence
length, readability, and sentiment, each book can be treated as a
point in a small feature space. Scaling these features and measuring
distance between books gives a simple way to rank which books are
most similar in style to a chosen one, the same underlying idea
behind KNN.

In [35]:
pip install scikit-learn scipy

  Using cached scipy-1.17.1-cp311-cp311-macosx_14_0_arm64.whl.metadata (62 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.3/8.3 MB 7.2 MB/s  0:00:01 eta 0:00:01
Using cached scipy-1.17.1-cp311-cp311-macosx_14_0_arm64.whl (20.3 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3/3 [scikit-learn] [scikit-learn]
Note: you may need to restart the kernel to use updated packages.


In [36]:
# scale style features so no single metric dominates the distance calculation
from sklearn.preprocessing import StandardScaler

style_features = ['type_token_ratio_sample', 'avg_sentence_length', 'flesch_kincaid_grade', 'avg_sentiment', 'sentiment_volatility']

scaler = StandardScaler()
scaled_values = scaler.fit_transform(df[style_features])

scaled_df = pd.DataFrame(scaled_values, columns=style_features)
scaled_df['book'] = df['book']

In [37]:
# compute style distance between every pair of books
from scipy.spatial.distance import pdist, squareform

distances = squareform(pdist(scaled_df[style_features]))
distance_df = pd.DataFrame(distances, index=scaled_df['book'], columns=scaled_df['book'])

In [38]:
def most_similar(book, n=3):
    return distance_df[book].sort_values().iloc[1:n + 1]

most_similar('austen-emma.txt')

book
austen-persuasion.txt    1.990302
carroll-alice.txt        2.833550
chesterton-ball.txt      2.852937
Name: austen-emma.txt, dtype: float64

In [39]:
distance_df.loc['austen-emma.txt', 'austen-sense.txt']

np.float64(2.9023492874486316)

Interestingly, Sense and Sensibility does not appear among Emma's
three closest matches, despite being written by the same author.
Checking the actual distance confirms this, Emma to Sense and
Sensibility is 2.902, only slightly further than Emma to Alice in
Wonderland (2.834) or The Ball and the Cross (2.853). Sense and
Sensibility's outlier sentiment volatility from the previous section
is likely responsible, pulling it just far enough away to edge out
shared authorship as the stronger signal. This shows how sensitive a
similarity ranking can be to a single outlier feature, even when the
overall difference in distance is small.